### PR TITLE
Makes discussion appear "READ" after being open

### PR DESCRIPTION
**Pull request by gabrielpoca**
The status property needs to be added to the DiscussionSerializer.

The problem here is that we are making two requests to get all information on
a discussion.

* The first one is on discussions/index, loading a small ammount of 
information for each discussions.

* The second one is for discussions/show. This requests ignores the cached
values and loads again all the information needed for a single discussion
(except the status).

Once someone opens a discussions/show a request is made to mark that page as
read. After that request we need to refresh the information we have on the
discussion. For this we call the discussions/show again (not the 
discussions/index). For this reason we need it to return status.

### CODE REVIEW

**Comment by gabrielpoca** [comment]
@naps62 I would appreciate feedback on this one, I don't like to duplicate the status method, do you have any suggestion?

**Comment by naps62** [comment]
I guess a better approach would be to use composition (maybe even use Concerns?), and move the `status` method to a module? The same should be probably be happening with EditableSerializer, it should be a mixin instead of a super class

**Comment by gabrielpoca** [comment]
@naps62 I was thinking the same thing, although I don't know what to call it

**Comment by gabrielpoca** [comment]
@naps62 what do you think?

File: app/serializers/concerns/subscribable.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module Subscribable
+  def status
+    Subscription.for(discussion_id: object.id, user_id: scope.id).state
```
Comment (zamith) [inline app/serializers/concerns/subscribable.rb]: There is mostly a naming issue here. This concern claims to make an object subscribable, but it only provides it's status, it does not allow to subscribe to it.

Moreover, it is coupled with it being a discussion, since the `for` method takes a `discussion_id`. Also, this method is called `status`, whereas in the subscription it's called `state`. That's not very consistent.

Comment (gabrielpoca) [inline app/serializers/concerns/subscribable.rb]: Like I'm saying in the other comments, I'm open for suggestions please

**Comment by gabrielpoca** [comment]
@zamith I can also leave it duplicated, in this case I don't think it's an issue. Either way, I have no idea on the best way to do this

**Comment by zamith** [comment]
@gabrielpoca I pushed a possible way to do it. One that is more flexible for the future, imo.

**Comment by gabrielpoca** [comment]
@zamith you introduced a lot of complexity to resolve the problem of having a one-line method repeated in two serializers.

Also, the existence of these two serializers are a premature optimisation for not sending all the information in one request, we don't know yet if it's going to be slow. This could be solved in different ways, for instance, we could have a resource "Comment" that gets called once you open a show, instead of having the information for a "Discussion" separated in two calls.

**Comment by gabrielpoca** [comment]
Either way, if @naps62 agrees I can go with it too.

**Comment by naps62** [comment]
This seems like over-engineering

We now would have two discussion serializers that don't even accept a discussion as the object (or if they do, the state will probably be `nil`. I dont know how the serializers work in this case)

We also end up with two different solutions for solving two very similar problems:

* `editable` We wanted this method in both serializers. For that we defined a superclass
* `state` We also want this on both serializers. We already have a superclass, so now we wrap everything in a decorator instead. This has the added problem that I already stated above

This seems like we're just using arbitrary design patterns everytime a new problem appears

And, in my opinion, both of them should be solved with composition (I'd go with Concerns actually, I actually like them)

TL;DR: No, I don't agree with this. At least not the way it is now

**Comment by zamith** [comment]
@naps62 It is a bit too much for this exact problem, but I think it opens the door for other usages of this.

Using it with a regular discussion might be an issue that would have to be fixed.

If what you are calling concerns are the mixins, that's inheritance, not composition.

This was an experiment, of using the decorator pattern, which I feel we could take more advantage of. In this particular instance, I have no problem with just dropping the changes altogether and have the method duplicated.

**Comment by gabrielpoca** [comment]
I actually think the problem here is the way we are designing it. We should have a resource for "Discussion" and another for "Comments", the whole point of having these two serializers is avoid sending the comments with the discussion. If we separate this we'll still the same amount of requests, and it will make more sense to use and be easier to understand.

**Comment by gabrielpoca** [comment]
This is also opening other issues, now, when we create a new discussion you make a first POST request, then you need to make a request to discussions/index in order fill the information needed to show on the page because the POST doesn't return all information needed on a discussion. I'm closing this pr, this doesn't make sense, I'm opening a new one we a solution that makes more sense on the long run and doesn't require such complexity.

**Comment by zamith** [comment]
@gabrielpoca I'm not sure I follow everything, but how is it possible that the POST to create a discussion does not return everything and the index does? The index is supposed to be sending the compact one, not the other way round. If there is something in the compact that is not in the other, then that's the problem.
